feat: add Cairo VM builtins documentation (#911)
cairo-book/cairo-book: src/ch204-02-builtins-list.md

# Builtins List

> In this section, we'll list the different builtins that
> exist and give a brief description of their purpose.
> The sub-sections that follows details how each builtin

Reviewer: ```suggestion
The following table lists the different builtins implemented in the CairoVM and gives a brief description of their purpose.
The sub-sections that follows details how each builtin
```

> The sub-sections that follows details how each builtin
> is working, its cells organization if any andreferences
> to their actual implementation in different Cairo VM
> projects.

Reviewer: ```suggestion
For each builtin, a specific section details how it works, its cells organization if any, and references their actual implementation in different implementations of the Cairo VM.
```

> Some topics related to the builtin might peak your interest,
> so external resources are provided for you to read.
> For example how the Pedersen hash function works.

Reviewer: ```suggestion
Additional resources related to the operation performed by the builtin are provided if relevant.
```

| [Bitwise][bitwise]       | Computes the bitwise AND, XOR and OR of two felts `a` and `b`. `a & b`, `a ^ b` and `a \| b`.                                          |
| [EC OP][ec_op]           | Performs Elliptic Curve OPerations - For two points on the STARK curve `P`, `Q` and a scalar `m`, computes `R = P + mQ`.               |
| [Keccak][keccak]         | Computes the new state `s'` after applying the 24 rounds of the keccak-f1600 block permutation on a given state `s`.                   |
| [Poseidon][poseidon]     | Computes the new state `s'` after applying the 91 rounds of the hades block permutation on a given state `s`.                          |
| [Range Check96][rc96]    | Verify that a felt `x` is within the bounds `[0, 2**96)`.                                                                              |
| [AddMod][add_mod]        | Arithmetic Circuit Support - Computes the modular addition `c` of two felts `a`, `b` by batches. `c ≡ a + b mod(p)`                    |
| [MulMod][mul_mod]        | Arithmetic Circuit Support - Computes the modular multiplication `c` of two felts `a`, `b` by batches. `c ≡ a * b mod(p)`              |

> | [Segment Arena][seg_are] | Manages the Cairo dictionaries Only used by Cairo >= 1.0                                                                               |

Reviewer: ```suggestion
| [Segment Arena][seg_are] | Manages the Cairo dictionaries. Not used in Cairo Zero.|
```

> | [Gas][gas]               | Manages the available gas during the run. Used by StarkNet to handle its gas usage and avoid DoS.                                      |

Reviewer: ```suggestion
| [Gas][gas]               | Manages the available gas during the run. Used by Starknet to handle its gas usage and avoid DoS.                                      |
```

> | [System][system]         | Manages the StarkNet syscalls & cheatcodes.                                                                                            |

Reviewer: ```suggestion
| [System][system]         | Manages the Starknet syscalls & cheatcodes.                                                                                            |
```
